### PR TITLE
[wasm][debugger] Showing "Frame not in module" after vscode-js-debug bump on VS

### LINKED ISSUE #86754: Frame not in module when hitting a breakpoint in Blazor WASM
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

After upgrading to Visual Studio 17.6.0 and 17.6.1 the debugger shows me a _Source Not Available_ page  with the body:

```
Frame not in module
The current stack frame was not found in a loaded module. Source cannot be shown for this location.
```

The curious thing is that in the .razor file the breakpoints selected before executing the application are kept without warnings, but if I add one while the application is running, it appears with the warning: "The breakpoint will not currently be hit. Unbound breakpoint".

When I open the debug window "Modules" I can check that the module "Application.server.dll" has the status "Symbols loaded" and that the "symbol file" is correct.

I have tried deleting the .obj and .bin directories, I have tried with .net 6 and .net 7 (with their corresponding updated NuGet packages), I have tried with different browsers, and in all of them, the error is the same.

When debugging no error appears in the Output window.

The curious thing is that if I create a new application with the wizard, there is no problem and Visual Studio stops at the line in the .razor file. I have analyzed both projects and I do not see what could be the cause of my main project causing this problem.

### Expected Behavior

- That Visual Studio allows me to add breakpoints without displaying the error "The breakpoint will not currently be hit. Unbound breakpoint".
- Visual Studio stops at the correct line in the .razor file once it hits the breakpoint.

### Steps To Reproduce

_No response_

### Exceptions (if any)

_No response_

### .NET Version

7.0.302

### Anything else?

- ASP.NET Core version: 6.0.16 or 7.0.5
- Visual Studio 17.6.1

```
.NET SDK:
 Version:   7.0.302
 Commit:    990cf98a27

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\7.0.302\

Host:
  Version:      7.0.5
  Architecture: x64
  Commit:       8042d61b17

.NET SDKs installed:
  3.1.426 [C:\Program Files\dotnet\sdk]
  5.0.408 [C:\Program Files\dotnet\sdk]
  6.0.311 [C:\Program Files\dotnet\sdk]
  7.0.302 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.29 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.24 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.28 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.29 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.23 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.17 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

### PR DESCRIPTION
**Pull request by thaystg**
In a blazor app with authentication enabled, 4 execution contexts are created, but the last two of them are created for the authentication thing and then they are destroyed when the login process is over. Once they are destroyed BrowserDebugProxy should not send scriptID's from this context when there is a Debugger.Pause, it should send the scriptId of the last not destroyed context.

Before this PR we were sending the scriptId of a context that was already destroyed and vscode-js-debug extension started to consider this information that was ignored before.

In my opinion we should revert vscode-js-debug change until we have versions that fixes it on .net6, .net7 and .net8.

fixes https://github.com/dotnet/runtime/issues/86754, https://github.com/dotnet/runtime/issues/87407
@joj @connor4312 @radical @danroth27

### CODE REVIEW

**Comment by connor4312** [comment]
Node.js also has a [similar problem](https://github.com/nodejs/node/issues/47438) that we have a [workaround for](https://github.com/microsoft/vscode-js-debug/blob/main/src/adapter/threads.ts#L1415-L1422). I'm okay putting in a similar workaround for Blazor until this gets fixed. Or maybe the debugger should just create execution context IDs if they don't exist and shrug its shoulder at implementors who "misbehave". This means scripts will not get cleared correctly on navigation or refresh, but it's better than breaking.

**Comment by radical** [comment]
cc @javiercn

**Comment by thaystg** [comment]
@radical can you please review?

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs [hunk 1]
```diff
@@ -45,24 +45,35 @@ public MonoProxy(ILogger logger, int runtimeId = 0, string loggerId = "", ProxyO
...
+            contexts[sessionId].Add(executionContext);
             return previous;
         }
 
+        internal bool TryGetExecutionContextValue(SessionId id, out ExecutionContext executionContext)
+        {
+            executionContext = null;
+            if (!contexts.TryGetValue(id, out List<ExecutionContext> contextList))
+                return false;
+            executionContext = contextList.Last<ExecutionContext>();
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:68]: `contextList` could be empty, and `Last` would then throw.

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs [hunk 2]
```diff
@@ -151,6 +162,13 @@ protected override async Task<bool> AcceptEvent(SessionId sessionId, JObject par
                         }
                         return true;
                     }
+                case "Runtime.executionContextDestroyed":
+                    {
+                        int id = args["executionContextId"].Value<int>();
+                        contexts.TryGetValue(sessionId, out var contextList);
+                        contextList.RemoveAll(x => x.Id == id);
```
Comment (radical) [inline src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:169]: If this makes `contextList` empty, then should it be removed?

**Comment by radical** [comment]
> We were sending the scriptId of a context that was already destroyed and vscode-js-debug extension started to consider this information that was ignored before.

When was this scriptId being sent? And how does that change with this PR?

File: src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs [hunk 3]
```diff
@@ -151,6 +162,13 @@ protected override async Task<bool> AcceptEvent(SessionId sessionId, JObject par
                         }
                         return true;
                     }
+                case "Runtime.executionContextDestroyed":
+                    {
+                        int id = args["executionContextId"].Value<int>();
+                        contexts.TryGetValue(sessionId, out var contextList);
```
[the remark this answers is not in the document]
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:168]: I don't think so, we need to send this message to vscode-js-debug.

[the remark this answers is not in the document]
Comment (thaystg) [reply src/mono/wasm/debugger/BrowserDebugProxy/MonoProxy.cs:168]: Ohhh, understood

**Comment by thaystg** [comment]
> > We were sending the scriptId of a context that was already destroyed and vscode-js-debug extension started to consider this information that was ignored before.
> 
> When was this scriptId being sent? And how does that change with this PR?

When we have a debugger.paused, the scriptId of the callstack is being sent.

As we were considering the CURRENT context the last one created, but this one was already destroyed, so we need to send the scriptID for the last not deleted context.

**Comment by radical** [comment]
> > > We were sending the scriptId of a context that was already destroyed and vscode-js-debug extension started to consider this information that was ignored before.
> > 
> > 
> > When was this scriptId being sent? And how does that change with this PR?
> 
> When we have a debugger.paused, the scriptId of the callstack is being sent.
> 
> As we were considering the CURRENT context the last one created, but this one was already destroyed, so we need to send the scriptID for the last not deleted context.

And how does this happen? Is the previous not deleted context still valid?

**Comment by thaystg** [comment]
> > > > We were sending the scriptId of a context that was already destroyed and vscode-js-debug extension started to consider this information that was ignored before.
> > > 
> > > 
> > > When was this scriptId being sent? And how does that change with this PR?
> > 
> > 
> > When we have a debugger.paused, the scriptId of the callstack is being sent.
> > As we were considering the CURRENT context the last one created, but this one was already destroyed, so we need to send the scriptID for the last not deleted context.
> 
> And how does this happen? Is the previous not deleted context still valid?

In a page that has OIDC authentication for example.
It creates context 1 2 3 and 4, destroys 3 and 4, and the 1 and 2 continues valid. So we need to send the scriptID of the context 2.

**Comment by radical** [comment]
Can you please add these details to the PR description?

**Review by radical — APPROVED** [review]

**Pull request merged**